data/tests/rulesets.cpp.in

```diff
@@ -0,0 +1,61 @@
+#include <QProcess>
+
+#include <QtTest>
+
+// Macro values replaced by cmake
+#define RULESETS "$<JOIN:${RULESET_LIST},", ">"
+#define SERVER_PATH "$<TARGET_FILE:freeciv21-server>"
+
+/**
+ * Ruleset-related tests
+ */
+class test_rulesets : public QObject {
+  Q_OBJECT
+
+private slots:
+  void load_data();
+  void load();
+};
+
+/**
+ * Generates test data for load()
+ */
+void test_rulesets::load_data()
+{
+  QTest::addColumn<QString>("name");
+  QTest::addColumn<bool>("exists");
+
+  const char *names[] = {"default", RULESETS};
+  for (auto &name : names) {
+    QTest::newRow(name) << name << true;
+  }
+
+  // A ruleset that doesn't exist
+  QTest::newRow("error") << "error" << false;
+}
+
+/**
+ * Tries to spawn a server with a ruleset.
+ */
+void test_rulesets::load()
+{
+  QFETCH(QString, name);
+  QFETCH(bool, exists);
+
+  QProcess p;
+  p.start(SERVER_PATH, {QStringLiteral("-r"), name});
```

Comment:
Does this fall back to the default ruleset on failure to load?

Reply:
Not anymore, the server now exits if any failing command in the startup script. This is checked in the `cli` test.